src/pm/hydra/proxy/pmip.c

```diff
@@ -149,40 +149,41 @@ int main(int argc, char **argv)
     /* collect the singleton's exit_status */
     if (PMIP_pg_0()->is_singleton) {
         struct pmip_pg *pg_0 = PMIP_pg_0();
-        HYDU_ASSERT(pg_0, status);
 
         HYDU_ASSERT(pg_0->num_procs == 1, status);
         HYDU_ASSERT(pg_0->downstreams[0].pid == HYD_pmcd_pmip.singleton_pid, status);
         /* We won't get the singleton's exit status. Assume it's 0. */
         if (pg_0->downstreams[0].exit_status == PMIP_EXIT_STATUS_UNSET) {
             pg_0->downstreams[0].exit_status = 0;
         }
-    } else {
-        /* Wait for the processes to finish */
-        int total_count = PMIP_get_total_process_count();
-        while (1) {
-            pid = waitpid(-1, &ret_status, 0);
-
-            /* Find the pid and mark it as complete. */
-            if (pid > 0) {
-                struct pmip_downstream *tmp = PMIP_find_downstream_by_pid(pid);
-                if (tmp) {
-                    if (tmp->exit_status == PMIP_EXIT_STATUS_UNSET) {
-                        tmp->exit_status = ret_status;
-                    }
-                    done++;
+
+        done++;
+    }
+
+    /* Wait for the processes to finish */
+    int total_count = PMIP_get_total_process_count();
+    while (done < total_count) {
+        pid = waitpid(-1, &ret_status, 0);
+
+        /* Find the pid and mark it as complete. */
+        if (pid > 0) {
+            struct pmip_downstream *tmp = PMIP_find_downstream_by_pid(pid);
+            if (tmp) {
+                if (tmp->exit_status == PMIP_EXIT_STATUS_UNSET) {
+                    tmp->exit_status = ret_status;
                 }
+                done++;
             }
+        }
 
-            /* If no more processes are pending, break out */
-            if (done == total_count)
-                break;
+        /* If no more processes are pending, break out */
+        if (done == total_count)
+            break;
```

Comment:
do we still need this break now that the loop condition is basically the same?

Reply:
It was meant to skip the `HYDT_dmx_wait_for_event(0)` but I guess it is harmless to have the extra poll anyway. I'll try remove this break later.

Reply:
Added fixup. Will squash before merge.